### PR TITLE
Handle coalesce partition handle when bucket count does not match but compatible

### PR DESCRIPTION
**Pull request by rongrong**
Should we have release note for this? It's technically a fix. Also do we have release note for default session value changes?

```
== RELEASE NOTES ==
General Changes
* Fix query failures with incompatible partition handle when session property ``partial_merge_pushdown_strategy`` is set to ``PUSH_THROUGH_LOW_MEMORY_OPERATORS`` and ``optimize_full_outer_join_with_coalesce`` is set to ``true`` and query has mismatched partition and uses ``FULL OUTER JOIN`` with ``COALESCE``.
```

### CODE REVIEW

**Comment by rongrong** [comment]
cc @caithagoras

File: presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java [hunk 1]
```diff
@@ -317,9 +317,9 @@ else if (argument instanceof VariableReferenceExpression) {
     }
 
     // Maps VariableReferenceExpression in both partitions to an COALESCE expression, keeps constant arguments unchanged.
-    public Optional<Partitioning> translateToCoalesce(Partitioning other)
+    public Optional<Partitioning> translateToCoalesce(Partitioning other, Metadata metadata, Session session)
     {
-        checkArgument(this.handle.equals(other.handle), "incompatible partitioning handles: %s != %s", this.handle, other.handle);
+        checkArgument(metadata.getCommonPartitioning(session, this.handle, other.handle).isPresent(), "incompatible partitioning handles: %s != %s", this.handle, other.handle);
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java]: Should we also update the error message?

**Comment by caithagoras** [comment]
Consider squashing the commits into one: You can add description about default value changes into the commit message of the 3rd commit.

**Comment by caithagoras** [comment]
We should have release notes for the fix. Something like

```
General Changes
* Fix query failure for ``CROSS JOIN`` with ``COALESCE`` when bucket counts are compatible but mismatched.
```

**Comment by caithagoras** [comment]
Thanks for the fix! lgtm, will let @wenleix take a look.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java (hunk 1 above)
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java]: what about:
```
    this.handle.equals(other.handle) || metadata.isRefinedPartitioningOver(session, this, other) || metadata.isRefinedPartitioningOver(session, other, this)
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java [hunk 2]
```diff
@@ -449,9 +449,12 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
                                 .build();
                     }
 
-                    if (isOptimizeFullOuterJoinWithCoalesce(session) && probeProperties.getNodePartitioning().isPresent() && buildProperties.getNodePartitioning().isPresent()) {
+                    if (isOptimizeFullOuterJoinWithCoalesce(session) &&
+                            probeProperties.getNodePartitioning().isPresent() &&
+                            buildProperties.getNodePartitioning().isPresent() &&
+                            metadata.getCommonPartitioning(session, probeProperties.getNodePartitioning().get().getHandle(), buildProperties.getNodePartitioning().get().getHandle()).isPresent()) {
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: what about 
```
   && metadata.isRefinedPartitioningOver(session, probeProperties.getNodePartitioning().get().getHandle(), buildProperties.getNodePartitioning().get().getHandle()).isPresent()) || metadata.isRefinedPartitioningOver()...
```

You might want to save `probeProperties.getNodePartitioning().get().getHandle()` and `buildProperties.getNodePartitioning().get().getHandle()).isPresent()` as local variable.

**Comment by wenleix** [comment]
`optimize_mismatched_bucket_count` cannot be default to true, this is because engine cannot reason whether the "compatible bucket" will result in more memory usage or not, for example:

```
          AGGR
            |
          JOIN
         /     \
 TS(2048)    TS(128)
```

Now it will be executed as if there are 128 buckets, and cause increase in local memory usage. This is the motivation to deprecate `optimize_mismatched_bucket_count` and introduce refined partitioning, so engine can reason why the partitioning is refined or coarsened .

**Pull request merged**